### PR TITLE
Started phase 1 adding IDO and sub-IDO classes

### PR DESCRIPTION
**Pull request by Mttbnchtt**
I added IDO and sub-IDO as subclasses of CGIAR objective specification. I will then add all the IDOs.

### CODE REVIEW

**Comment by Mttbnchtt** [comment]
I have just started, but reviews are welcomed at any point.

**Comment by pbuttigieg** [comment]
Hi @Mttbnchtt - the label "class created to address issue" should only be used for released classes. It's clear that should be retitled (I'll do that now) as it is confusing.

File: src/sdgio-edit.owl [hunk 1]
```diff
@@ -1515,6 +1519,24 @@ AnnotationAssertion(<http://www.geneontology.org/formats/oboInOwl#hasExactSynony
...
 SubClassOf(<http://purl.unep.org/sdg/SDGIO_00020232> <http://purl.unep.org/sdg/SDGIO_00000251>)
 
+# Class: <http://purl.unep.org/sdg/SDGIO_00020236> (intermediate development outcome)
+
+AnnotationAssertion(<http://purl.org/dc/elements/1.1/creator> <http://purl.unep.org/sdg/SDGIO_00020236> <http://orcid.org/0000-0062-3253-4387>)
+AnnotationAssertion(<http://purl.org/dc/elements/1.1/date> <http://purl.unep.org/sdg/SDGIO_00020236> "2019-04-28T11:19:21Z"^^xsd:dateTime)
+AnnotationAssertion(<http://www.geneontology.org/formats/oboInOwl#hasExactSynonym> <http://purl.unep.org/sdg/SDGIO_00020236> "IDO")
+AnnotationAssertion(Annotation(<http://www.geneontology.org/formats/oboInOwl#hasDbXref> <https://hdl.handle.net/10947/3865>) rdfs:comment <http://purl.unep.org/sdg/SDGIO_00020236> "A CGIAR objective specification which 1) names an objective the achievement of which contributes to the achievement of a specific system level outcome and 2) indicates to which system level outcome a piece of CGIAR research contributes and how such a piece of research contributes to that system level outcome.")
```
Comment (pbuttigieg) [inline src/sdgio-edit.owl]: Some definitional points:
* It's a bit strange to say that something is an objective and also names an objective.
* Are the IDOs limited to research activity? I think operational processes fall under this, but perhaps @marieALaporte can provide insight here.
* It seems that the IDO definition is very overloaded: it does far too much for a single objective specification. It can't both direct "what" is to be done and "how" it is to be done. The "how" must be defined in a `plan specification`. @Mttbnchtt: if and IDO is really supposed to do all this, we will have to unpack this concept into both ObjSpec and a PlanSpec.

I'd rephrase this definition to:

```
A CGIAR objective specification which describes the intended endpoint and outcome(s) of a process which, itself, is part of and contributes to a larger process directed by a system level outcome.

File: src/sdgio-edit.owl [hunk 2]
```diff
@@ -1515,6 +1519,24 @@ AnnotationAssertion(<http://www.geneontology.org/formats/oboInOwl#hasExactSynony
...
+AnnotationAssertion(rdfs:label <http://purl.unep.org/sdg/SDGIO_00020236> "intermediate development outcome"@en)
+SubClassOf(<http://purl.unep.org/sdg/SDGIO_00020236> <http://purl.unep.org/sdg/SDGIO_00000251>)
+
+# Class: <http://purl.unep.org/sdg/SDGIO_00020237> (sub-intermediate development outcome)
```
Comment (pbuttigieg) [inline src/sdgio-edit.owl]: The label here is a bit off, I think. 
It should be "sub-outcome of an intermediate development outcome"

File: src/sdgio-edit.owl [hunk 3]
```diff
@@ -1515,6 +1519,24 @@ AnnotationAssertion(<http://www.geneontology.org/formats/oboInOwl#hasExactSynony
...
+
+AnnotationAssertion(Annotation(<http://www.geneontology.org/formats/oboInOwl#hasDbXref> <https://hdl.handle.net/10947/3865>) <http://purl.obolibrary.org/obo/IAO_0000115> <http://purl.unep.org/sdg/SDGIO_00020237> "A CGIAR objective specification which 1) names an objective the achievement of which contributes to the achievement of an IDO and 2) names an objective of a specific and limited activity of researchers and policy makers.")
```
Comment (pbuttigieg) [inline src/sdgio-edit.owl]: Similar definitional issues to the IDO def. Suggested rephrasing:

```
A CGIAR objective specification which describes the intended endpoint and outcome(s) of a process which, itself, is part of and contributes to a larger process directed by an intermediate level outcome.
```

File: src/sdgio-edit.owl [hunk 4]
```diff
@@ -2727,6 +2749,22 @@ AnnotationAssertion(rdfs:comment <http://purl.unep.org/sdg/SDGIO_00020235> "Mult
 AnnotationAssertion(rdfs:label <http://purl.unep.org/sdg/SDGIO_00020235> "improvement of natural resources and ecosystem services system level outcome"@en)
 ClassAssertion(<http://purl.unep.org/sdg/SDGIO_00020232> <http://purl.unep.org/sdg/SDGIO_00020235>)
 
+# Individual: <http://purl.unep.org/sdg/SDGIO_00020238> (Increased resilience of the poor to climate change and other shocks)
```
Comment (pbuttigieg) [inline src/sdgio-edit.owl]: Remove all capital letters that are not part of proper nouns.

File: src/sdgio-edit.owl [hunk 5]
```diff
@@ -2727,6 +2749,22 @@ AnnotationAssertion(rdfs:comment <http://purl.unep.org/sdg/SDGIO_00020235> "Mult
...
+
+AnnotationAssertion(<http://purl.org/dc/elements/1.1/creator> <http://purl.unep.org/sdg/SDGIO_00020238> <http://orcid.org/0000-0062-3253-4387>)
+AnnotationAssertion(<http://purl.org/dc/elements/1.1/date> <http://purl.unep.org/sdg/SDGIO_00020238> "2019-04-28T12:16:49Z"^^xsd:dateTime)
+AnnotationAssertion(<http://www.geneontology.org/formats/oboInOwl#hasExactSynonym> <http://purl.unep.org/sdg/SDGIO_00020238> "1"^^xsd:integer)
```
Comment (pbuttigieg) [inline src/sdgio-edit.owl]: "1" is not a valid synonym. I think we should have a DbXref with a CURIE-style ID. 
Something like CGIAR-IDO:1.

@Mttbnchtt: Are there no alternate IDs from the various APIs we were informed about?

File: src/sdgio-edit.owl [hunk 6]
```diff
@@ -2727,6 +2749,22 @@ AnnotationAssertion(rdfs:comment <http://purl.unep.org/sdg/SDGIO_00020235> "Mult
...
+AnnotationAssertion(rdfs:label <http://purl.unep.org/sdg/SDGIO_00020238> "Increased resilience of the poor to climate change and other shocks"@en)
+ClassAssertion(<http://purl.unep.org/sdg/SDGIO_00020236> <http://purl.unep.org/sdg/SDGIO_00020238>)
+
+# Individual: <http://purl.unep.org/sdg/SDGIO_00020239> (Enhanced smallholder market access)
```
Comment (pbuttigieg) [inline src/sdgio-edit.owl]: Replace capital letters with lowercase

File: src/sdgio-edit.owl [hunk 7]
```diff
@@ -2727,6 +2749,22 @@ AnnotationAssertion(rdfs:comment <http://purl.unep.org/sdg/SDGIO_00020235> "Mult
...
+
+AnnotationAssertion(<http://purl.org/dc/elements/1.1/creator> <http://purl.unep.org/sdg/SDGIO_00020239> <http://orcid.org/0000-0062-3253-4387>)
+AnnotationAssertion(<http://purl.org/dc/elements/1.1/date> <http://purl.unep.org/sdg/SDGIO_00020239> "2019-04-28T12:19:53Z"^^xsd:dateTime)
+AnnotationAssertion(<http://www.geneontology.org/formats/oboInOwl#hasExactSynonym> <http://purl.unep.org/sdg/SDGIO_00020239> "2"^^xsd:integer)
```
Comment (pbuttigieg) [inline src/sdgio-edit.owl]: As with SDGIO_00020238, replace synonym with DbXref in CURIE style.

File: src/sdgio-edit.owl (hunk 1 above)
Comment (Mttbnchtt) [reply src/sdgio-edit.owl]: I like how you rephrase your definition and I will implement it. I will also try to de-load the definition. Here is the main piece of text from which I got the definition: "To ground these SLOs at the level of research activities, CGIAR has introduced the concept of Intermediate Development Outcome (IDOs), which enable researchers to think through the contexts in which their outputs might contribute to development outcomes" (p. 14 of the document titled _CGIAR Strategy and Results Framework_).

On a minor point, I took (and still take) an objective _specification_ to be a linguistic entity whose designatum is an objective. This is why I began my suggested definition as "an objective specification which names etc.". However, your suggestions "objective speciication which describes ..." is surely clearer and I will adopt it.

Comment (Mttbnchtt) [reply src/sdgio-edit.owl]: On second thought, I am not sure that an objective specification describes something rather than being a label to refer to something. Like 'point' refers to a point but does not describe a point (or the concept of point). Therefore, I will proceed to de-load the definition but, for now, I will not say that an objective specification describes something. If you confirm your suggestion of saying that an objective specification describes something, I will do so as soon as I read your new comment.

(Also let me know if I am just being picky in an unhelpful way or observations of this type are useful. Building an ontology is partly an art and I am trying to understand your way of doing it.)

Comment (Mttbnchtt) [reply src/sdgio-edit.owl]: For now I have modified the definition of IDO as follows:
A CGIAR objective specification which names an objective the achievement of which contributes to the achievement of a specific system level outcome.

Moreover, I modified the definition of sub-IDO as follows:
A CGIAR objective specification which names an objective the achievement of which contributes to the achievement of an intermediate development outcome.

As I said, I am pausing on your suggestion to say that an IDO describes something. However, I am pausing only because I would like to hear from you more about this point. I will adopt your suggestion if you confirm it.

File: src/sdgio-edit.owl (hunk 2 above)
Comment (Mttbnchtt) [reply src/sdgio-edit.owl]: I understand that my label sounds odd (to be gentle). However, this is exactly what CGIAR people say. I used this piece of text: "Below this level [i.e. IDOs] are Sub-Intermediate Development Outcome (sub-IDOs)" (p. 14 of the document _CGIAR Strategy and Results Framework)_. 

It may be relevant to note that the language of the document is sometimes not that great. Do you think that the need for a decently-formulated label is more important than faithfulness to the source of that label? How do you reason in cases like this one?

File: src/sdgio-edit.owl (hunk 3 above)
Comment (Mttbnchtt) [reply src/sdgio-edit.owl]: I de-loaded the definition of sub-IDO. However, as before, I ask you to confirm suggestion of saying that a sub-IDO _describes_ something rather than _refers_ (or names or something like that) to something.

File: src/sdgio-edit.owl (hunk 4 above)
Comment (Mttbnchtt) [reply src/sdgio-edit.owl]: Done. 
Just to understand, several names of instances begin with capital letters even if the words are not proper nouns. Is there a general rule? (I got a screenshot that I can share by email. Apparently, I cannot attach files here.)

File: src/sdgio-edit.owl (hunk 6 above)
Comment (Mttbnchtt) [reply src/sdgio-edit.owl]: Done. Thanks for noticing all these important things.

File: src/sdgio-edit.owl (hunk 5 above)
Comment (Mttbnchtt) [reply src/sdgio-edit.owl]: I added the DbXref CGIAR-IDO:1 as an annotation of the label. 
I was unsure about where to put the DbXref, since here we do not have definitions. When we created the SLOs we put DbXref as annotation of the definition. It seems to me that a label  plays a role that is somehow analogous to the role of a definition when there is no definition. So, I added the DbXref as annotation of the label. Let me know if this is wrong. 

Moreover, I do not know what a CURIE-style ID is. I googled the expression but got no information. I simply copied and pasted the ID that you wrote. If it is important for me to know, could you please refer me to a place where I can find more information?

I added the DbXref as an IRI, as last time. Let me know if this is appropriate in this case too. Thanks very much.

File: src/sdgio-edit.owl (hunk 7 above)
Comment (Mttbnchtt) [reply src/sdgio-edit.owl]: Done, with the caveat explain in the comment above.

File: src/sdgio-edit.owl (hunk 1 above)
Comment (Mttbnchtt) [reply src/sdgio-edit.owl]: To make matters more complicated, the json file does call IDOs descriptions. Here is a snippet:

  {
    "code": "1",
    "description": "Increased resilience of the poor to climate change and other shocks",
    "isCrossCutting": false,
    "srfCrossCuttingIssue": null
  }

So maybe your suggestion is right (because a description describes something). Let me know what you think is appropriate. Let me also know if it is inappropriate for me to do all this thinking about a single issue that is probably a minor one.

Comment (Mttbnchtt) [reply src/sdgio-edit.owl]: In the end, I accepted Pier Luigi's suggestion, as I understood that he repeated it on Skype. We will talk about this on 29 Apr 2019.

File: src/sdgio-edit.owl (hunk 3 above)
Comment (Mttbnchtt) [reply src/sdgio-edit.owl]: In the end, I accepted Pier Luigi's suggestion, as I understood that he repeated it on Skype. We will talk about this on 29 Apr 2019.

File: src/sdgio-edit.owl (hunk 2 above)
Comment (marieALaporte) [reply src/sdgio-edit.owl]: If we want to stick with the CGIAR language, it would be better to keep the main labels as they are stated in the SRF document. The label that @pbuttigieg proposed, which is going to be more understandable by external people, can be added as exact synonym. What do you think?

File: src/sdgio-edit.owl (hunk 1 above)
Comment (marieALaporte) [reply src/sdgio-edit.owl]: > To make matters more complicated, the json file does call IDOs descriptions. Here is a snippet:

I think that in the json output when it says description it should be just saying label or something similar. No real description are really given. You can see the same pattern for other calls. Let me confirm this and come back to you @Mttbnchtt

Comment (marieALaporte) [reply src/sdgio-edit.owl]: > Are the IDOs limited to research activity? I think operational processes fall under this, but perhaps @marieALaporte can provide insight here.

@pbuttigieg I am going to look into that too as I am not sure about it

File: src/sdgio-edit.owl (hunk 5 above)
Comment (pbuttigieg) [reply src/sdgio-edit.owl]: > I was unsure about where to put the DbXref, since here we do not have definitions.

In this case, it's a DbXref on a class level. In the SLO case, the xref was for the definition, not the whole class.

> It seems to me that a label plays a role that is somehow analogous to the role of a definition when there is no definition. 

No, the label is just a label. For instances, there sometimes isn't a definition in an ontological sense - the class does that work. 

> So, I added the DbXref as annotation of the label. Let me know if this is wrong.

This should be on the class level, not on the label. Generally, you add APs to other APs when they are solely relevant to them, rather than the whole class.

> Moreover, I do not know what a CURIE-style ID is.

It's an OBO convention and follows the NAMESPACE:ID format, like [ENVO:00001000] or [SDGIO:00000000]. 

This old ENVO issue may be informative: https://github.com/EnvironmentOntology/envo/issues/699
http://owlcollab.github.io/oboformat/doc/obo-syntax.html#2.6

@cmungall do we still have to escape the brackets for these?

> I added the DbXref as an IRI, as last time. Let me know if this is appropriate in this case too

As long as it's resolvable, this should work.

File: src/sdgio-edit.owl (hunk 7 above)
Comment (pbuttigieg) [reply src/sdgio-edit.owl]: Use the IRIs provided by @marieALaporte in #142 and then resolve this comment.

File: src/sdgio-edit.owl (hunk 1 above)
Comment (pbuttigieg) [reply src/sdgio-edit.owl]: @Mttbnchtt: we use "describes" to align with IAO: 
http://purl.obolibrary.org/obo/IAO_0000005

This is where conversations on the metaphysics should go (e.g. see https://github.com/information-artifact-ontology/IAO/issues/213). They are useful, but not likely to be consequential in our application scenario and will just slow things down. We can realign if there's a need to later or if there's a philosophical breakthrough (which can take far too long for our use case). 
>>To make matters more complicated, the json file does call IDOs descriptions. Here is a snippet:
>
> I think that in the json output when it says description it should be just saying label or something similar. No real description are really given.

Yes, this looks like something that's just a label for the format and nothing that should be thought about too ontologically.

File: src/sdgio-edit.owl (hunk 2 above)
Comment (Mttbnchtt) [reply src/sdgio-edit.owl]: Following Pier Luigi's suggestion, repeated in our Skype conversation on 29 Apr 2019, I used "sub-outcome of an intermediate development outcome" as the label, for the sake of linguistic correctness. I added "sub-intermediate development outcome" and "sub-IDO" as two distinct hasExactSynonym.

File: src/sdgio-edit.owl (hunk 1 above)
Comment (Mttbnchtt) [reply src/sdgio-edit.owl]: got it

File: src/sdgio-edit.owl (hunk 2 above)
Comment (pbuttigieg) [reply src/sdgio-edit.owl]: There's no such thing as "sub-intermediate", I'm going to insist on labels that make sense. Future editing of the ontology requires consistency, otherwise we'll have to explain every internal organisations logic to other users.

> The label that @pbuttigieg proposed, which is going to be more understandable by external people, can be added as exact synonym.

Keep in mind, that SDGIO is external to the CGIAR and seeks to externalise it's strategy. The outward facing label should take precedence.

That being said, if there's a special label needed by an organisation, it can either be added as a synonym or as its own AP. For example, we have "UNEP preferred label" so that UN Environment's systems can display a label that they prefer. 

Practically speaking, systems should be focused on the IRIs and not the labels anyway.

File: src/sdgio-edit.owl (hunk 4 above)
Comment (pbuttigieg) [reply src/sdgio-edit.owl]: That shouldn't be the case, could you add a few examples here?
In some cases (e.g. targets, indicators), the instance labels are essentially sentences, which follow sentence grammar.

File: src/sdgio-edit.owl (hunk 1 above)
Comment (Mttbnchtt) [reply src/sdgio-edit.owl]: I am going to resolve this conversation.
Summary of the issue: the definition of the IDO class was too complicated and did not align with conceptual rules states in IAO.
Solution: I modified the definition according to Pier Luigi's suggestion.

File: src/sdgio-edit.owl (hunk 2 above)
Comment (Mttbnchtt) [reply src/sdgio-edit.owl]: I am going to resolve this conversation.
Summary of the issue: the phrase "sub-intermediate development outcome" is hardly understandable by a user external to CGIAR.
Solution: I modified the label according to Pier Luigi's suggestion.

File: src/sdgio-edit.owl (hunk 3 above)
Comment (Mttbnchtt) [reply src/sdgio-edit.owl]: I am going to resolve this conversation.
Summary of the issue: the definition of the sub-IDO class was too complicated and did not align with conceptual rules states in IAO.
Solution: I modified the definition according to Pier Luigi's suggestion.

File: src/sdgio-edit.owl (hunk 5 above)
Comment (Mttbnchtt) [reply src/sdgio-edit.owl]: I am going to resolve this conversation.
Summary of the issue: the xrefs of the IDOs was invalid.
Solution: I modified the xrefs using the URLs provided by @marieALaporte  in #142.
